### PR TITLE
establish_channel_with_system

### PR DESCRIPTION
**Pull request by xlc**
Implements https://github.com/polkadot-fellows/RFCs/issues/82
Allow any parachain to have bidirectional channel with any system parachains

Looking for initial feedbacks before continue finish it

TODOs:
- [x] docs
- [x] benchmarks
- [x] tests

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5552257

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5552222

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5552239

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5552240

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5552256

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5552241

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 1]
```diff
@@ -836,6 +836,34 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// TODO write docs
```
Comment (ggwpez) [inline polkadot/runtime/parachains/src/hrmp.rs:840]: Just FYI we have a CI job that scans for the string `FAIL-CI` to remind of todos 😄

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 2]
```diff
@@ -836,6 +836,34 @@ pub mod pallet {
...
+			origin: OriginFor<T>,
+			recipient: ParaId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_parachain(<T as Config>::RuntimeOrigin::from(origin))?;
+
+			ensure!(recipient.is_system(), Error::<T>::ChannelCreationNotAuthorized);
+
+			let config = <configuration::Pallet<T>>::config();
+			let max_message_size = config.hrmp_channel_max_message_size;
+			let max_capacity = config.hrmp_channel_max_capacity;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:853]: I thought we discussed lowering this to some "standard" size/capacity?

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 3]
```diff
@@ -836,6 +836,34 @@ pub mod pallet {
...
+
+			Self::init_open_channel(sender, recipient, max_capacity, max_message_size)?;
+			Self::accept_open_channel(recipient, sender)?;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:856]: Most chains would want the system to be able to send them messages too, right? But not want to do a referendum to open a channel. So this should repeat with the reverse.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5633480

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 4]
```diff
@@ -270,6 +270,10 @@ pub mod pallet {
 		/// implementation should be the same as `Balance` as used in the `Configuration`.
 		type Currency: ReservableCurrency<Self::AccountId>;
 
+		/// The default channel size and capacity to use when opening a channel to a system
+		/// parachain.
+		type DefaultChannelSizeAndCapacityToSystem: Get<(u32, u32)>;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:275]: I would add an `integrity_test` that these are below the maximum allowed by `Configuration`.

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 5]
```diff
@@ -836,6 +840,36 @@ pub mod pallet {
...
 			Ok(())
 		}
+
+		/// TODO write docs
+		#[pallet::call_index(10)]
+		#[pallet::weight(<T as Config>::WeightInfo::establish_system_channel())] // TODO benchmarks
+		pub fn establish_channel_with_system(
+			origin: OriginFor<T>,
+			recipient: ParaId,
+		) -> DispatchResultWithPostInfo {
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:850]: Since it's bidirectional, the system chain is not just a recipient. I would change this to either `interlocutor` or `target_system_chain`.

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 4 above)
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:275]:
```suggestion
		type DefaultChannelSizeAndCapacityWithSystem: Get<(u32, u32)>;
```

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 6]
```diff
@@ -836,6 +840,36 @@ pub mod pallet {
...
+			let sender = ensure_parachain(<T as Config>::RuntimeOrigin::from(origin))?;
+
+			ensure!(recipient.is_system(), Error::<T>::ChannelCreationNotAuthorized);
+
+			let (max_message_size, max_capacity) = T::DefaultChannelSizeAndCapacityToSystem::get();
+
+			// create bidirectional channell
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:857]:
```suggestion
			// create bidirectional channel
```

**Review by joepetrowski — COMMENTED** [review]
I think the approach is OK, just a few naming comments.

**Review by acatangiu — COMMENTED** [review]
approach looks good, will approve once PR is final (todos, tests, benchmarks, etc)

**Review by kianenigma — COMMENTED** [review]
LGTM

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 7]
```diff
@@ -1004,7 +1038,7 @@ impl<T: Config> Pallet<T> {
 		loop {
 			// bail if we've iterated over all items.
 			if idx == 0 {
-				break
+				break;
```
Comment (ggwpez) [inline polkadot/runtime/parachains/src/hrmp.rs:1041]: no unrelated changes please, `cargo +nightly-2023-11-01 fmt` it is.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5653225

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 7 above)
Comment (xlc) [reply polkadot/runtime/parachains/src/hrmp.rs:1049]: still having those semi. not sure why. will just manually delete them

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5654892

File: polkadot/runtime/rococo/src/weights/runtime_parachains_hrmp.rs [hunk 8]
```diff
@@ -331,4 +331,14 @@ impl<T: frame_system::Config> runtime_parachains::hrmp::WeightInfo for WeightInf
 			.saturating_add(T::DbWeight::get().reads(1))
 			.saturating_add(T::DbWeight::get().writes(1))
 	}
+	fn establish_channel_with_system() -> Weight {
+		// Proof Size summary in bytes:
+		//  Measured:  `417`
+		//  Estimated: `6357`
+		// Minimum execution time: 629_674_000 picoseconds.
+		Weight::from_parts(640_174_000, 0)
```
Comment (xlc) [inline polkadot/runtime/rococo/src/weights/runtime_parachains_hrmp.rs:339]: copied from establish_system_channel

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 4 above)
Comment (xlc) [reply polkadot/runtime/parachains/src/hrmp.rs]: config is from storages so it is not really possible to assert it at build time.
if the value is above the max defined in config, the call will fail, and I think that's fine.

Comment (joepetrowski) [reply polkadot/runtime/parachains/src/hrmp.rs]: Ah right 👍

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 9]
```diff
@@ -270,6 +274,10 @@ pub mod pallet {
 		/// implementation should be the same as `Balance` as used in the `Configuration`.
 		type Currency: ReservableCurrency<Self::AccountId>;
 
+		/// The default channel size and capacity to use when opening a channel to a system
+		/// parachain.
+		type DefaultChannelSizeAndCapacityWithSystem: Get<(u32, u32)>;
```
Comment (ggwpez) [inline polkadot/runtime/parachains/src/hrmp.rs:279]: Can you please mention the exact values that this holds? So `proposed_max_message_size` and `proposed_max_capacity`.

**Review by ggwpez — APPROVED** [review]

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 4 above)
Comment (ggwpez) [reply polkadot/runtime/parachains/src/hrmp.rs]: We do run `integrity_test` in externalities, so at least the genesis values can be checked.

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 10]
```diff
@@ -270,6 +274,10 @@ pub mod pallet {
 		/// implementation should be the same as `Balance` as used in the `Configuration`.
 		type Currency: ReservableCurrency<Self::AccountId>;
 
+		/// The default channel size and capacity to use when opening a channel to a system
+		/// parachain.
```
Comment (ggwpez) [inline polkadot/runtime/parachains/src/hrmp.rs:278]: Does it need metadata with `pallet::constant`?

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 9 above)
Comment (xlc) [reply polkadot/runtime/parachains/src/hrmp.rs:279]: you mean the value that will be used in the relaychain runtime? I don’t exactly have a number in mind but if I have to provide one, I will just make it half of the max value

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 10 above)
Comment (xlc) [reply polkadot/runtime/parachains/src/hrmp.rs:278]: we can do it. but given the max value are not constant, do we want to make those constant? maybe we want to use the parameter pallet to specify the value?

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 9 above)
Comment (ggwpez) [reply polkadot/runtime/parachains/src/hrmp.rs:279]: Yea that is fine as a comment as well i guess. like "good values for these would be half of the maximum".

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 10 above)
Comment (ggwpez) [reply polkadot/runtime/parachains/src/hrmp.rs:278]: Ah yea, i am often not sure what to put in metadata 😅

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 9 above)
Comment (xlc) [reply polkadot/runtime/parachains/src/hrmp.rs:279]: we haven’t had discussions about what the value will be and I don’t want to imply anything here. I am deferring the decision until the PR to the relaychain runtime to upgrade polkadot-sdk

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 4 above)
Comment (xlc) [reply polkadot/runtime/parachains/src/hrmp.rs]: but why check with genesis values?

**Comment by xlc** [comment]
this is ready for final review

File: polkadot/runtime/parachains/src/hrmp.rs (hunk 4 above)
Comment (ggwpez) [reply polkadot/runtime/parachains/src/hrmp.rs]: Because they should also be correct? but its not a must, just an idea.

File: polkadot/runtime/parachains/src/hrmp.rs [hunk 11]
```diff
@@ -836,6 +844,50 @@ pub mod pallet {
...
+				T::DefaultChannelSizeAndCapacityWithSystem::get();
+
+			// create bidirectional channel
+			Self::init_open_channel(sender, target_system_chain, max_capacity, max_message_size)?;
+			Self::accept_open_channel(target_system_chain, sender)?;
+
+			Self::init_open_channel(target_system_chain, sender, max_capacity, max_message_size)?;
+			Self::accept_open_channel(sender, target_system_chain)?;
+
+			Self::deposit_event(Event::HrmpSystemChannelOpened {
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp.rs:875]: The docs on this event say between two system chains. You should update them to re-use the event.

File: polkadot/runtime/parachains/src/hrmp/benchmarking.rs [hunk 12]
```diff
@@ -513,6 +513,21 @@ mod benchmarks {
...
+		let recipient_id: ParaId = 2u32.into();
+
+		let sender_origin: crate::Origin = sender_id.into();
+
+		// make sure para is registered, and has zero balance.
+		register_parachain_with_balance::<T>(sender_id.into(), Zero::zero());
+		register_parachain_with_balance::<T>(recipient_id, Zero::zero());
+
+		#[extrinsic_call]
+		_(sender_origin, recipient_id);
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp/benchmarking.rs:528]: Should have some assertions to make sure it worked.

File: prdoc/pr_3721.prdoc [hunk 13]
```diff
@@ -0,0 +1,12 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: New call `hrmp.establish_channel_with_system` to allow parachains to establish a channel with the system parachain
```
Comment (joepetrowski) [inline prdoc/pr_3721.prdoc:4]:
```suggestion
title: New call `hrmp.establish_channel_with_system` to allow parachains to establish a channel with a system parachain
```

**Review by joepetrowski — APPROVED** [review]

**Comment by xlc** [comment]
not sure why PRDoc check failed?

**Comment by acatangiu** [comment]
> not sure why PRDoc check failed?

You need to specify what kind of bump should the affected crates get: https://github.com/paritytech/polkadot-sdk/blob/665e3654ceca5a34e8ada66a9805fa7b76fc9ebb/prdoc/pr_3927.prdoc#L13

File: polkadot/runtime/westend/src/lib.rs [hunk 14]
```diff
@@ -1203,11 +1203,16 @@ impl pallet_message_queue::Config for Runtime {
 
 impl parachains_dmp::Config for Runtime {}
 
+parameter_types! {
+	pub const DefaultChannelSizeAndCapacityWithSystem: (u32, u32) = (4096, 4);
```
Comment (acatangiu) [inline polkadot/runtime/westend/src/lib.rs:1207]: why is this so small for westend? (rococo gets `(51200, 500)`)

**Review by acatangiu — APPROVED** [review]

File: prdoc/pr_3721.prdoc [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+title: New call `hrmp.establish_channel_with_system` to allow parachains to establish a channel with a system parachain
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      This PR adds a new call `hrmp.establish_channel_with_system` that allows a parachain origin to open a bidirectional channel with a system parachain.
+
+crates: 
+  - name: polkadot-runtime-parachains
```
Comment (acatangiu) [inline prdoc/pr_3721.prdoc:12]:
```suggestion
- name: polkadot-runtime-parachains
  bump: minor
```

File: polkadot/runtime/westend/src/lib.rs (hunk 14 above)
Comment (xlc) [reply polkadot/runtime/westend/src/lib.rs:1207]: https://github.com/paritytech/polkadot-sdk/blob/212179cb990d3155463324ca59e8ed5d8ed82fc1/polkadot/node/service/src/chain_spec.rs#L144-L145

otherwise the benchmarks won't work

File: substrate/frame/support/src/traits/randomness.rs [hunk 16]
```diff
@@ -40,7 +40,7 @@ pub trait Randomness<Output, BlockNumber> {
 	/// Get the basic random seed.
 	///
 	/// In general you won't want to use this, but rather `Self::random` which allows
-	/// you to give a subject for the random result and whose value will be
+	/// you to give a subject for the random result and whose value will eb
```
Comment (joepetrowski) [inline substrate/frame/support/src/traits/randomness.rs:43]:
```suggestion
	/// you to give a subject for the random result and whose value will be
```

File: polkadot/runtime/parachains/src/hrmp/benchmarking.rs [hunk 17]
```diff
@@ -103,24 +110,24 @@ where
 	));
 
 	if matches!(until, ParachainSetupStep::Requested) {
-		return output
+		return output;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp/benchmarking.rs:113]:
```suggestion
		return output
```

File: polkadot/runtime/parachains/src/hrmp/benchmarking.rs [hunk 18]
```diff
@@ -103,24 +110,24 @@ where
...
 	}
 
 	assert_ok!(Hrmp::<T>::hrmp_accept_open_channel(recipient_origin.into(), sender));
 	if matches!(until, ParachainSetupStep::Accepted) {
-		return output
+		return output;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp/benchmarking.rs:118]:
```suggestion
		return output
```

File: polkadot/runtime/parachains/src/hrmp/benchmarking.rs [hunk 19]
```diff
@@ -103,24 +110,24 @@ where
...
 	}
 
 	Hrmp::<T>::process_hrmp_open_channel_requests(&configuration::ActiveConfig::<T>::get());
 	if matches!(until, ParachainSetupStep::Established) {
-		return output
+		return output;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp/benchmarking.rs:123]:
```suggestion
		return output
```

File: polkadot/runtime/parachains/src/hrmp/benchmarking.rs [hunk 20]
```diff
@@ -103,24 +110,24 @@ where
...
 	}
 
 	let channel_id = HrmpChannelId { sender, recipient };
 	assert_ok!(Hrmp::<T>::hrmp_close_channel(sender_origin.clone().into(), channel_id));
 	if matches!(until, ParachainSetupStep::CloseRequested) {
 		// NOTE: this is just for expressiveness, otherwise the if-statement is 100% useless.
-		return output
+		return output;
```
Comment (joepetrowski) [inline polkadot/runtime/parachains/src/hrmp/benchmarking.rs:130]:
```suggestion
		return output
```

**Review by joepetrowski — APPROVED** [review]
